### ISSUE TITLE
Too many open files error (file leak when creating runners/communicators)

### ISSUE THREAD

**Issue by giovannipizzi** [aiidateam/aiida-core#2386]
When running tests, I get `[Errno 24] Too many open files` on Mac. 
This can be temporarily worked around using `ulimit -n 1024` to increase the limit to 1024 (as in linux) instead of the Mac OS X limit of 256. However, this pointed out a leak in the current code.

By a bit of inspection (printing the number of open FDs with `psutil.Process().num_fds()`) I realised that the source of the problem can be tracked down as follows:
- call to `run_get_node` (in process_function.py)
- this calls `self.create_runner(with_persistence=False)` 
  - internally, this goes down to kiwipi to the `RmqThreadCommunicator.connect` class method, that creates a new loop as this is not passed as a kwargs: `loop = loop or ioloop.IOLoop()` (this opens 3 FDs)
  - the following call to `loop.run_sync(lambda: connection_factory(**connection_params))` in the same method opens one more FD
  - when returning a runner (`return runners.Runner(**settings)` in `create_runner()`, file `manager.py`, as we don't pass a loop, a new loop is created in the __init__ of `Runner`: 3 more FDs
- occasionally, `proc = process_class(inputs=inputs, runner=runner)` inside `run_get_node` will clean up many processes - but this does not seem to be consistent, only happen in 1 of the 8 tests of `db.work.calcfunctions`, not clear why... also in other places of the code sometimes some FDs disappear but it's not clear why.
- the call to `runner.close()` in `run_get_node()` does not clean up any of these open FDs (only one eventually gets cleaned up of all the open FDs)

I think this might be due to the default creation of loops when entering functions such as `.connect()`, `create_runner`, or in the communicator, that are never properly stopped and closed.

I imagine this might even occur in production runs and not only on the test environment, even if I didn't debug this (neither I debugged if this happens also on Linux and not only on Mac).

Note: #1923 maybe was related to this?

**Comment by borellim** [comment]
I just saw this on a "production" environment, on Ubuntu 16.04, with aiida-core develop (6e9711046753332933f982971db1d7ac7e7ade58). Restarting the daemon solves the issue.

I got the following stats from `lsof` (first column is number of open files, second is the truncated process name; files open by multiple processes are counted twice):
```
7288 pool
5676 gmain
5676 gdbus
4616 dconf\x20
2553 verdi
1289 evolution
1257 python2
877 indicator
578 compiz
314 unity-pan
```

The only change after restarting the daemon is:
```
1682 verdi
```

The script:
```
lsof | awk '{map[$1]+=1} END{for (i in map) print map[i], i}' | sort -h -r | head -n10
```

**Comment by ltalirz** [comment]
Thanks @borellim for the script!

@danieleongari just encountered this issue in production calculations again - the runners failed with the `Too many open files` error when opening input files etc.
```
# Initial
23265 code
18069 ThreadPoo
13523 Composito
11745 verdi
11377 gmain
11365 gdbus
10524 JS\x20Hel
7026 dconf\x20
6924 TaskSched
6873 slack

# after closing vscode
15510 ThreadPoo
11745 verdi
11616 JS\x20Hel
10972 Composito
9967 gmain
9955 gdbus
6924 TaskSched
6871 slack
6464 dconf\x20
4574 ZMQbg/1

# after verdi daemon restart
15789 ThreadPoo
11904 JS\x20Hel
11052 Composito
10278 gmain
10266 gdbus
6989 slack
6924 TaskSched
6774 dconf\x20
4574 ZMQbg/1
4031 pool
```

He writes 
> but after restarting the daemon when I run `verdi process play --all` the process are unreachable

Note that  `code` (here: the binary from visual studio code) has even more open files
There is a bug report on this on the vscode github but they claim this comes from 3rd party components: https://github.com/microsoft/vscode/issues/53965

**Comment by sphuber** [comment]
> but after restarting the daemon when I run verdi process play --all the process are unreachable

This is normal, at least directly after restarting the daemon. When the daemon restarts, it will have to reaccept the tasks from RabbitMQ and reconstruct the process in memory. As long as this process is not completed, it is technically "not reachable". Depending on how many active processes you have, it will take a little while before all processes are active again and they can be resumed (think 10s of seconds to few minutes maximum). If, however, even after waiting, some processes remain unreachable, that would most likely be due to a bug.

**Comment by danieleongari** [comment]
@sphuber, good point. I just checked and the process are still unreachable after more than one hour.

**Comment by sphuber** [comment]
@danieleongari are all processes unreachable, or just a subset of them? Can you check the daemon log if you see any occurrences of the string `failed to recreate the process instance in order to continue it` or `Cannot continue process`? Can you also try to stop and start the daemon once more to see if that changes things. Finally, do you have the RabbitMQ admin panel installed by any chance? If so, we can check whether the process tasks actually still exist in the queue and are just not getting to the daemon or if the tasks no longer exist. In the case of the latter, I can give you a simple snippet to recreate them and "resurrect" the process, but for that we first need to confirm the tasks no longer exist.

**Comment by danieleongari** [comment]
> are all processes unreachable, or just a subset of them?

All, they are 14

> Can you check the daemon log if you see any occurrences of the string failed to recreate the process instance in order to continue it or Cannot continue process

None of these two strings are in the `.aiida/daemon/log/aiida-{profile}.log` file. The last message was that the daemon was restarted.

> Can you also try to stop and start the daemon once more to see if that changes things.

Nothing changed: I tried immediately after to resubmit the processes and after some minutes as you mentioned in your previous message.

> do you have the RabbitMQ admin panel installed by any chance? If so, we can check whether the process tasks actually still exist in the queue and are just not getting to the daemon or if the tasks no longer exist.

I managed to visualize this 
![image](https://user-images.githubusercontent.com/25899092/89207740-837f9a80-d5bb-11ea-938d-ebfc5f24754b.png)

Is this telling that no tasks are still existing? You can reach me on slack for a faster interaction.

**Comment by unkcpz** [comment]
I can't reproduce this issue in my computer, neither by running all test case or by running a real world workflow using `run_get_node`.  When the pytest is done or the process is finished, the number of open files of `lsof` command outputs is unchanged. 

> only happen in 1 of the 8 tests of `db.work.calcfunctions` 

where are these test cases locate? Maybe I can reproduce the issue there. @giovannipizzi 

I am in the latest develop branch of aiida_core, and also test this in the asyncio version.

**Comment by sphuber** [comment]
Just for reference, we checked with @danieleongari and there were indeed no process tasks in the corresponding process queue. The reason why is difficult to ascertain at distance. Although it _may_ have to do with the max number of file limit being reached, it is by far not the only possibility. And I find it a bit weird that _all_ of the tasks were gone and not just a few.

**Comment by danieleongari** [comment]
The problem occurred once again today:
* now the error in the report of the calculations is `OSError: [Errno 16] Device or resource busy` (for all cases)
* it is not related to a particular plugin because the same happened with `RasspaCalculation`s and `Cp2kCalculation`s
* the `process list` shows: `Pausing after failed transport task: update_calculation failed 5 times consecutively`
* `verdi process play` is uneffective: the calculation shows `Waiting for transport task: upload` for some minutes and then comes back to the original failure
* the only two considerable changes in my setup are (1) the update to `aiida-core v1.3`, (2) the yearly maintenance of EFL's clusters in July (I'm using Fidis computer) 
* I did not restart the deamon yet, but I'm afraid that doing so the processes get unreachable like the previous time

**Comment by greschd** [comment]
> 
>     * now the error in the report of the calculations is `OSError: [Errno 16] Device or resource busy` (for all cases)

At least when I test it on my machine, the error when opening too many files is `OSError: [Errno 24] Too many open files`. This seems like a slightly different error, though it could of course have the same root cause.

Do you happen to have some more traceback in the daemon log - s.t. we could figure out which operation the daemon was trying to do when it failed?

Also, how is the storage on your "local" (wherever the daemon runs) and remote (where the code runs) machines set up? Is it a local disk, NFS, Lustre, ...? Depending on which it is, the different error modes may be more or less likely.

**Comment by ltalirz** [comment]
Ok, just a few more pieces of info:

Global file open limit is far from being reached (27k open files, 3M limit)
```
$ cat /proc/sys/fs/file-nr
27584	0	3271698
```

File open limit for the `verdi` process (1024) is indeed reached
```
$ cat /proc/24253/limits | grep open
Max open files            1024                 4096                 files
$ lsof -p 24253 | grep -P "^(\w+\s+){3}\d+\D+" | wc -l
1024
```
(see [stackoverflow](https://serverfault.com/a/964752/321020) for more explanation)

I attach the output of `lsof -p 24253 | grep -P "^(\w+\s+){3}\d+\D+" | wc -l` (i.e. the files opened by `verdi` that actually count towards the file open limit): [lsof-verdi-counting.txt](https://github.com/aiidateam/aiida-core/files/5027471/lsof-verdi-counting.txt)

Except for a few at the beginning, they are all triples of the form
```
verdi   24253 daniele 1021r     FIFO    0,12      0t0  6690390 pipe
verdi   24253 daniele 1022w     FIFO    0,12      0t0  6690390 pipe
verdi   24253 daniele 1023u  a_inode    0,13        0    13699 [eventpoll]
```

**Comment by ltalirz** [comment]
@greschd Currently, it seems like the calculations are failing at
```
$ verdi process report XXX
...
 |   File "<string>", line 4, in raise_exc_info
 |   File "/home/daniele/aiida1/aiida_core/aiida/engine/transports.py", line 88, in do_open
 |   File "/home/daniele/aiida1/aiida_core/aiida/transports/plugins/ssh.py", line 416, in open
 |   File "/home/daniele/anaconda3/envs/aiida1/lib/python3.6/site-packages/paramiko/client.py", line 340, in connect
 |   File "/home/daniele/anaconda3/envs/aiida1/lib/python3.6/site-packages/paramiko/client.py", line 204, in _families_and_addresses
 |   File "/home/daniele/anaconda3/envs/aiida1/lib/python3.6/socket.py", line 745, in getaddrinfo
 | OSError: [Errno 16] Device or resource busy
```
i.e. I guess when trying to open a connection via paramiko.

However, I remember that last time it was indeed when trying to open some random file (which may explain the difference in the error message).

The root cause, however, seems to be the fact that the `verdi` worker process has maxed out the number of open files.

P.S. The main daemon process (`verdi daemon start-circus 1`) has only 29 open files

**Comment by ltalirz** [comment]
This is a bit of a shot in the dark (not sure whether this is related), but I came across [this SO post](https://stackoverflow.com/a/55665722/1069467) on pyzmq opening large numbers of sockets.

It seems this happens only with the pyzmq wheel installed via `pip` (manylinux build), which uses sockets for inter-thread communication, but not if it is built locally or installed via `conda`.
Running the test script provided in the answer confirms that we have the "wheel" version which opens many extra sockets.
While I guess this may not be the root cause of the issue, perhaps it could somehow exacerbate it?

In any case, here is the full output of `lsof` for further future analysis: [lsof.zip](https://github.com/aiidateam/aiida-core/files/5027645/lsof.zip)

**Comment by ltalirz** [comment]
We've temporarily increased the file open limit per process via `ulimit -n unlimited`; this results in a limit of ~1M:
```
ulimit -a  | grep desc
-n: file descriptors                1048576
```

Interestingly, now that the daemon has been restarted and all calculations have been replayed, the number of open file descriptors for the `verdi run_daemon` process is actually quite low:
```
$ lsof -p 12996 | grep -P "^(\w+\s+){3}\d+\D+" | wc -l
135
```
Full output here: [lsof-verdi-running.txt](https://github.com/aiidateam/aiida-core/files/5027839/lsof-verdi-running.txt)

Does this mean there is a spike when creating/submitting the calculations?

In any case, we'll keep monitoring this number.

**Comment by sphuber** [comment]
> Does this mean there is a spike when creating/submitting the calculations?

Make sure that you check the process tasks actually still exist within RabbitMQ. @danieleongari should know how to do this now as I showed him a few days back. He reported a case where all processes stopped were unreachable, and we checked and verified that the tasks were gone completely indeed. Most likely, active processes that are with a daemon runner will most likely also keep a file open, for communication between the worker and RabbitMQ, for example for the heartbeat. If there are no process tasks, there will also be no communication and so no files. This may be an (partial) explanation why you see less files after the restart. Note again that the process tasks disappearing should never happen and is anomolous behavior. I haven't been able to track down where or why this is happening in Daniele's case.

**Comment by danieleongari** [comment]
@sphuber in this new case of today there was not anymore the issue of lost processes. 
I simply run:
```
ulimit -n unlimited
verdi daemon restart
verdi process play --all
```
and all the calculations and work chains proceeded as expected.

**Comment by ltalirz** [comment]
Just an update on the monitoring of the open file handles: While the number of open handles does go down by one from time to time, overall there clearly does seem to be a leak:
![verdi-monitor](https://user-images.githubusercontent.com/1053245/89464816-5aa10600-d771-11ea-898a-d10885925454.png)
Perhaps @danieleongari can comment what happened around 21:00, probably some new calculations kicking in.

**Comment by ltalirz** [comment]
During today's AiiDA meeting, @ramirezfranciscof mentioned he encountered the issue with aiida-core 1.3.0 as well.
It was mentioned that this may not be in the aiida-core code base but in the dependencies. Here is what changed between aiida 1.2.1 (where this issue has not been observed to my knowledge) and 1.3.0:

 - click-config-file: 0.5.0 => 0.6.0
 - kiwipy[rmq]~=0.5.1 => kiwipy[rmq]~=0.5.5
 - plumpy~=0.14.5 => plumpy~=0.15.0
 - new tqdm dependency

given that the leaks are of the `[eventpoll]` type as mentioned [above](https://github.com/aiidateam/aiida-core/issues/2386#issuecomment-669092595), this raises the question whether changes in plumpy or kiwipy could be responsible for this.
Any ideas @muhrin @unkcpz ?

**Comment by greschd** [comment]
Note that it could also be new releases of packages, or second-order dependencies, that just so happen to be installed with the "newer" 1.3 installs.

Unless we can get a complete `pip freeze` of both a leaking and a non-leaking variant, I wouldn't restrict the search to the dependencies listed above.

**Comment by danieleongari** [comment]
Thanks to @ltalirz, we are now using a workaround to increase the limit on the number of open files for all the daemon workers:
```
# increase open file limit for workers
worker_pids=$(ps -ef  | grep "devel run_daemon" | grep $1 | awk 'BEGIN { ORS=" " } { print $2}')

for pid in ${worker_pids}; do
  cmd="prlimit --pid $pid --nofile=16384:16384"
  echo "Running: $cmd"
  $cmd
  new_limit=`cat /proc/${pid}/limits | grep open | awk '{print $4}'`
  echo "New file open limit for process $pid : $new_limit"
done
```
where the `$1` argument is your aiida profile.

Note that without this, it's impossible for me to work with AiiDA, as I continuously incur in transport failure.

**Comment by ltalirz** [comment]
Since @unkcpz asked for a minimal working example in order to reproduce the issue, here it comes:

 * This [branch](https://github.com/ltalirz/aiida-core/tree/test-filedescriptor-issue) adds logging of open file descriptors to `/tmp/fds` to the `MultiplyAddWorkChain` 
 * Here is an example submit script
```python
from aiida.engine import submit
from aiida.orm import Int, load_code
from aiida.plugins import WorkflowFactory

builder = WorkflowFactory('arithmetic.multiply_add').get_builder()

builder.x = Int(1)
builder.y = Int(2)
builder.z = Int(3)
# Code was setup code using
# verdi computer setup -L tutor -H localhost -T local -S direct -w `echo $PWD/work` -n
# verdi computer configure local tutor --safe-interval 5 -n
# verdi code setup -L add --on-computer --computer=tutor -P arithmetic.add --remote-abs-path=/bin/bash -n
builder.code = load_code('add@tutor')

future = submit(builder)
print(future)
```

You can follow the file descriptors via `tail -f /tmp/fds` , or you check `cat /proc/${pid}/limits | grep open | awk '{print $4}'` of the daemon worker (only on linux; if you know the cmd on macos feel free to edit my answer)

**Comment by unkcpz** [comment]
@ltalirz Thanks! that really helpful. By running this example in asyncio aiida_core https://github.com/aiidateam/aiida-core/pull/4218, I find the file descriptor remain consistent. I locate the difference is that I use the global runner when launching the process function. Surely, there are some fds opened and not properly closed in the event loop which concomitant create with the local runner (maybe the communicator or the persister). Anyway, by explicitly close the local runner's event loop solve the issue, see PR https://github.com/aiidateam/aiida-core/pull/4307 which free the open fds of the loop of local runner.

**Comment by sphuber** [comment]
Fixed in PR #4307
